ledger/ledger.go

```diff
@@ -40,73 +45,145 @@ type Ledger interface {
 	RootHash() string
 	// GetPreviousValue executes a get against a previous version of the ledger, using that version's root hash.
 	GetPreviousValue(previousRootHash, key string) (result string, err error)
+	// EraseRootHash re-claims any memory used by this version of history, preserving bits shared with other versions.
+	EraseRootHash(rootHash string) error
+	// Stats gives basic storage info regarding the ledger's underlying cache
+	Stats() cache.Stats
+	// GetAll returns the entire state of the ledger
+	GetAll() (map[string]string, error)
+	// GetAllPrevious returns the entire state of the ledger at an arbitrary version
+	GetAllPrevious(string) (map[string]string, error)
 }
 
 type smtLedger struct {
-	tree *smt
+	tree    *smt
+	history *history
+	// this map will grow forever.  how do we limit it?
+	keyCache byteCache
 }
 
 // Make returns a Ledger which will retain previous nodes after they are deleted.
-func Make(retention time.Duration) Ledger {
-	return smtLedger{tree: newSMT(hasher, nil, retention)}
+// the retention parameter has been removed in favor of EraseRootHash, but is left
+// here for backwards compatibility
+func Make(_ time.Duration) Ledger {
+	return &smtLedger{
+		tree:    newSMT(hasher, nil),
```

Comment:
nit: it seems newSMT makes `cache.NewTTL(forever, time.Second)`. This seems like a waste of a goroutine and it polls every 1s for something that will never happen (assuming `forever` doesn't occur :slightly_smiling_face: ). Its not made worse by this PR though as far as I can tell.

Reply:
This was useful previously when we were setting expiration for non-current nodes, but the Erase() function makes that unnecessary.  Honestly, we don't need this to be an expiring cache at all now, but this library doesn't have a simple cache, and when it was written Martin was requiring that all cache usages in istio use this library.  Not sure why.

I can move the evictionInterval to a day if that helps...

Reply:
Moving to a day doesn't help, I think that would either lead to no benefits or lead to obscure issues at a day?

maybe we can just tweak it to not run the goroutine if there is no ttl set

Reply:
Oh, yeah, we can just set it to 0.  Will do.

https://github.com/istio/pkg/blob/master/cache/ttlCache.go#L99